text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+  receiver. If this connection fails because the `ArrowStreamService` is not implemented by the target, the exporter
+  must automatically fall back on the behavior of the classic OTLP protocol. A configuration parameter could be added to
+  disable this default behavior.
+
+The implementation of these two rules should allow a seamless and
+adaptive integration of OTLP Arrow into the current ecosystem
+generally.
+
+For the prototype specifically, which is a fork of the OpenTelemetry
+collector codebase, we have derived the OTLP/gRPC-Arrow exporter and
+receiver as set of changes directly to the `receiver/otlpreceiver` and
+`exporter/otlpexporter` components, with new `internal/arrow` packages
+in both.  With every collector release we merge the OTLP-Arrow changes
+with the mainline components to maintain this promise of
+compatibility.
+
+OTLP-Arrow supports conveying the gRPC metadta (i.e., http2 headers) using a dedicated `bytes` field.  Metadata is
+encoded using [hpack](https://datatracker.ietf.org/doc/rfc7541/) like a typical unary gRPC request.
+
+Specifically:
+
+#### OTLP/gRPC Receiver
+
+When Arrow is enabled, the OTLP receiver listens for both the standard unary gRPC service OTLP and OTLP-Arrow stream
+services.  Each stream uses an instance of the OTel-Arrow-Adapter's
+[Consumer](https://pkg.go.dev/github.com/f5/otel-arrow-adapter@v0.0.0-20230112224802-dafb6df21c97/pkg/otel/arrow_record#Consumer). Sets
+`client.Metadata` in the Context.
+
+#### OTLP/gRPC Exporter
+
+When Arrow is enabled, the OTLP exporter starts a fixed number of streams and repeatedly sends one `plog.Logs`,
+`ptrace.Traces`, or `pmetric.Metrics` item per stream request.  The `exporterhelper` callback first tries to get an
+available stream, blocking when none are available (or until the connection is downgraded), and then falls back to the
+standard unary gRPC path.  The stream-sending mechanism internally handles retries when failures are caused by streams
+restarting, while honoring the caller's context deadline, to avoid delays introduced by allowing these retries to go
+through the `exporterhelper` mechanism.
+
+Each stream uses an instance of the OTel-Arrow-Adapter's
+[Producer](https://pkg.go.dev/github.com/f5/otel-arrow-adapter@v0.0.0-20230112224802-dafb6df21c97/pkg/otel/arrow_record#Producer).
+
+When a stream fails specifically because the server does not recognize the Arrow service, it will not restart.  When all
+streams have failed in this manner, the connection downgrades by closing a channel, at which point the exporter behaves
+exactly as the standard OTLP exporter.
+
+The mechanism as described is vulnerable to partial failure scenarios.  When some of the streams are succeeding but some
+have failed with Arrow unsupported, the collector performance will be degraded because callers are blocked waiting for
+available streams.  The exact signal used to signal that Arrow and downgrade mechanism is seen as an area for future
```

Comment:
This for example can happen in rolling upgrade scenarios when load balancer sends client connections to receivers of different versions. For safety and simplicity perhaps just require that any single failure of ArrowStreamService should result in the whole exporter downgrading to OTLP unary.

Reply:
This sounds easy to change. https://github.com/open-telemetry/experimental-arrow-collector/issues/37

Reply:
I believe this has been mitigated. User have the option to disable downgrade entirely, or to set keepalive parameters that will recycle streams and thus ensure partial-downgrade is short-lived. I am leaving the issue open-telemetry/experimental-arrow-collector#37 open to capture other solutions, since I have observed that the code cannot _simply_ be changed to downgrade on the first failure.